File: data/src/main/java/com/popularmovies/vpaliy/data/mapper/Mapper.java
package com.popularmovies.vpaliy.data.mapper;


import java.util.ArrayList;
import java.util.List;

/**
 * Maps one a fake entity into a real
 * @param <To> real entity
 * @param <From> fake entity
 */
public abstract class Mapper<To, From> {
    public abstract To map(From from);
    public abstract From reverseMap(To to);

    public List<To> map(List<From> froms){
        if(froms!=null){
            List<To> result=new ArrayList<>(froms.size());
            for(From from:froms) result.add(map(from));
            return result;
        }
        return null;
    }

    public List<From> reverseMap(List<To> toList){
        if(toList!=null){
            List<From> result=new ArrayList<>(toList.size());
            for(To to:toList) result.add(reverseMap(to));
            return result;
        }
        return null;
    }
}

File: data/src/main/kotlin/com/popularmovies/vpaliy/data/mapper/Mapper.kt
package com.popularmovies.vpaliy.data.mapper

interface Mapper<Real, Fake> {
  fun map(list: List<Fake>): List<Real> {
    val result = arrayListOf<Real>()
    list.forEach {
      result.add(map(it))
    }
    return result
  }

  fun reverse(list: List<Real>): List<Fake> {
    val result = arrayListOf<Fake>()
    list.forEach {
      result.add(reverse(it))
    }
    return result
  }

  fun reverse(real: Real): Fake
  fun map(fake: Fake): Real
}